### PR TITLE
userd: handle help urls which requires prepending XDG_DATA_DIRS

### PR DESCRIPTION
**Pull request by kenvandine**
Pass through help urls to xdg-open as well as prepend $SNAP/usr/share to XDG_DATA_DIRS which is required to allow yelp to find the help files.

This fixes opening application help in desktop snaps that utilize yelp for user documentation.

### CODE REVIEW

File: userd/launcher.go [hunk 1]
```diff
@@ -97,7 +98,23 @@ func (s *Launcher) OpenURL(addr string) *dbus.Error {
...
 
-	if err = exec.Command("xdg-open", addr).Run(); err != nil {
+
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	xdg_data_dirs := []string{}
+	xdg_data_dirs = append(xdg_data_dirs, fmt.Sprintf("/snap/%s/current/usr/share", snap))
```
Comment (zyga) [inline userd/launcher.go]: The path is incorrect. Please use `dirs.SnapMountDir` instead.

File: userd/launcher.go [hunk 2]
```diff
@@ -97,7 +99,22 @@ func (s *Launcher) OpenURL(addr string) *dbus.Error {
...
 	}
 
-	if err = exec.Command("xdg-open", addr).Run(); err != nil {
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	xdg_data_dirs := []string{}
+	xdg_data_dirs = append(xdg_data_dirs, fmt.Sprintf("/%s/%s/current/usr/share", snap, dirs.SnapMountDir))
```
Comment (zyga) [inline userd/launcher.go]: Wait, isn't that the other way around? 

How about:
```suggestion
	xdg_data_dirs = append(xdg_data_dirs, filepath.Join(dirs.SnapMountDir, snap, "current/usr/share"))
```

Comment (kenvandine) [reply userd/launcher.go]: Doh!  Of course

File: .travis.yml [hunk 3]
```diff
@@ -41,6 +41,8 @@ jobs:
         apt:
           packages:
             python-launchpadlib
+      git:
```
Comment (zyga) [inline .travis.yml]: Can you propose this separately please?

**Comment by zyga** [comment]
This is failing on spread tests:
```
-----
+ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/pkgdb.sh
++ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/quiet.sh
+ . /home/gopath/src/github.com/snapcore/snapd/tests/lib/dirs.sh
++ export SNAP_MOUNT_DIR=/snap
++ SNAP_MOUNT_DIR=/snap
++ export LIBEXECDIR=/usr/lib
++ LIBEXECDIR=/usr/lib
++ export MEDIA_DIR=/media
++ MEDIA_DIR=/media
++ case "$SPREAD_SYSTEM" in
+ dbus-launch
++ xargs
+ export DBUS_SESSION_BUS_ADDRESS=unix:abstract=/tmp/dbus-kBSlEbTRZR,guid=e2d051cf4ee4a8acfe5be2295c14bc99 DBUS_SESSION_BUS_PID=26067
+ DBUS_SESSION_BUS_ADDRESS=unix:abstract=/tmp/dbus-kBSlEbTRZR,guid=e2d051cf4ee4a8acfe5be2295c14bc99
+ DBUS_SESSION_BUS_PID=26067
+ ping_launcher
+ dbus-send --session --dest=io.snapcraft.Launcher --type=method_call --print-reply / org.freedesktop.DBus.Peer.Ping
method return time=1544862875.485183 sender=:1.1 -> destination=:1.0 serial=4 reply_serial=2
+ cat
+ chmod +x /tmp/xdg-open
+ '[' -e /usr/bin/xdg-open ']'
+ mv /usr/bin/xdg-open /usr/bin/xdg-open.orig
+ touch /usr/bin/xdg-open
+ mount --bind /tmp/xdg-open /usr/bin/xdg-open
+ ensure_xdg_open_output https://snapcraft.io
+ rm -f /tmp/xdg-open-output
+ export DBUS_SESSION_BUS_ADDRESS=unix:abstract=/tmp/dbus-kBSlEbTRZR,guid=e2d051cf4ee4a8acfe5be2295c14bc99
+ DBUS_SESSION_BUS_ADDRESS=unix:abstract=/tmp/dbus-kBSlEbTRZR,guid=e2d051cf4ee4a8acfe5be2295c14bc99
+ chroot /snap/core/current /usr/bin/xdg-open https://snapcraft.io
user-open error: cannot find snap for connection: cannot find a snap for pid 26087
```

**Comment by jhenstridge** [comment]
One other potential roadbump: Yelp runs single-instance via GtkApplication.

If Yelp is already running when you attempt to open the help: URI, then the custom XDG_DATA_DIRS will likely be ignored, since all the new yelp process will do is notice the existing one and ask it to open the URI.

**Comment by pedronis** [comment]
So the spread tests are still failing, that needs addressing

**Comment by mvo5** [comment]
Looking at the spread failure it seems like the issue is that the spread test is not doing enough, i.e. it is not issuing a dbus call to trigger the xdg-open and because of that the snapFromSender fails with an error that it cannot find the snap for the sender.

**Comment by pedronis** [comment]
@kenvandine the spread test here needs attention

**Comment by kenvandine** [comment]
> @kenvandine the spread test here needs attention

Yeah, @mvo5 said he would help figure out how we can make the spread tests handle this.

**Comment by bboozzoo** [comment]
I'll take a look into why it's failing.

**Comment by bboozzoo** [comment]
Ok, so the spread `tests/main/xdg-open` should now pass. There is still a problem with CLA check.

@kenvandine can you reset the author of the commits that are signed with `kvandine@gnome.org` and try to drop https://github.com/snapcore/snapd/pull/6252/commits/c7d97e0cacc2b40a8d51cd46f5668ff05f403c31 ?

**Comment by bboozzoo** [comment]
The `tests/main/xdg-open-compat` test still fails. There is a couple of issues with the test:
- snapd-xdg-open uses a hardcoded list of supported schemes, so `help:` won't work
- the test currently executes the same thing as `tests/main/xdg-open` one, because `xdg-open` in the core snap uses the new interface only
- the test needs to be updated in the same way as `tests/main/xdg-open` was, so we actually need to call the interface from a running snap
- it didn't test what it advertises since ~March 2018

I'll try to get it fixed.

**Comment by bboozzoo** [comment]
The checks are all green now. Probably needs someone to take another look at the changes now.

**Comment by kenvandine** [comment]
> Ok, so the spread `tests/main/xdg-open` should now pass. There is still a problem with CLA check.
> 
> @kenvandine can you reset the author of the commits that are signed with `kvandine@gnome.org` and try to drop [c7d97e0](https://github.com/snapcore/snapd/commit/c7d97e0cacc2b40a8d51cd46f5668ff05f403c31) ?

I've rewritten the author for those but now there are conflicts.  I'm wondering if that is something I did or caused by upstream changes.

**Comment by mvo5** [comment]
One of the force pushed that happend to this repo made things seriously unhappy and we now have a 10k commit diff and tons of conflicts. I tried to cherry-pick based on the history of this PR in https://github.com/snapcore/snapd/pull/6493 and it seems that worked so I close this PR now.

**Pull request closed without merging**